File: app/src/main/java/com/gdgomaha/githubbrowser/home/detail/RepoDetailPresenter.java
package com.gdgomaha.githubbrowser.home.detail;

import com.gdgomaha.githubbrowser.data.RepoRepository;
import com.gdgomaha.githubbrowser.di.ForScreen;
import com.gdgomaha.githubbrowser.di.ScreenScope;
import com.gdgomaha.githubbrowser.utils.DisposableManager;

import javax.inject.Inject;
import javax.inject.Named;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;

@ScreenScope final class RepoDetailPresenter {

    @Inject RepoDetailPresenter(
            @Named("owner") String owner,
            @Named("repo") String repoName,
            RepoRepository repoRepository,
            @ForScreen DisposableManager disposableManager,
            RepoDetailViewModel viewModel
    ) {
        disposableManager.add(repoRepository.getRepo(owner, repoName)
                .observeOn(AndroidSchedulers.mainThread())
                .doOnSuccess(viewModel::processRepo)
                .observeOn(Schedulers.io())
                .flatMap(repo -> repoRepository.getContributors(repo.getContributorsUrl()))
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(viewModel::processContributors));
    }
}

File: app/src/main/java/com/gdgomaha/githubbrowser/home/detail/RepoDetailPresenter.kt
package com.gdgomaha.githubbrowser.home.detail

import com.gdgomaha.githubbrowser.data.RepoRepository
import com.gdgomaha.githubbrowser.di.ForScreen
import com.gdgomaha.githubbrowser.di.ScreenScope
import com.gdgomaha.githubbrowser.model.Contributor
import com.gdgomaha.githubbrowser.utils.DisposableManager
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import javax.inject.Inject
import javax.inject.Named

@ScreenScope internal class RepoDetailPresenter @Inject constructor(
        @Named("owner") owner: String,
        @Named("repo") repoName: String,
        repoRepository: RepoRepository,
        @ForScreen disposableManager: DisposableManager,
        viewModel: RepoDetailViewModel
) {

    init {
        disposableManager.add(repoRepository.getRepo(owner, repoName)
                .observeOn(AndroidSchedulers.mainThread())
                .doOnSuccess({ viewModel.processRepo(it) })
                .observeOn(Schedulers.io())
                .flatMap<List<Contributor>> { (_, _, _, _, _, _, contributorsUrl) -> repoRepository.getContributors(contributorsUrl) }
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ contributors, error -> viewModel.processContributors(contributors, error) }))
    }
}
